### PR TITLE
Improve snippet client management.

### LINKED ISSUE #55: Snippet client reference is not removed from ad object after teardown
'load_snippet' sets a snippet client reference attribute on the AndroidDevice object.
We should remove the attribute after snippet is stopped.

### LINKED ISSUE #56: load_snippet allows registering different packages under the same alias.
We should probably throw an error if attempts to register a different package under an existing alias is made?
```
ad.load_snippet('snippet', 'my.package')
ad.load_snippet('snippet', 'my.other.package') # Raise error
```

### PR DESCRIPTION
**Pull request by xpconanfan**
* Better handling of attempts to load snippet with duplicate info.
* Remove client attributes in stop_services.
* Add unit tests
fixes #55 
fixes #56

### CODE REVIEW

File: mobly/controllers/android_device.py [hunk 1]
```diff
@@ -509,26 +518,49 @@ def root_adb(self):
...
+                raise SnippetError(
+                    'Snippet package "%s" has already been loaded under name "%s".'
+                    % (package, client_name))
         host_port = utils.get_available_host_port()
         # TODO(adorokhine): Don't assume that a free host-side port is free on
         # the device as well. Both sides should allocate a unique port.
         device_port = host_port
         client = snippet_client.SnippetClient(
             package=package, port=host_port, adb_proxy=self.adb)
         self._start_jsonrpc_client(client, host_port, device_port)
```
Comment (dthkao) [inline mobly/controllers/android_device.py]: why a nested dict? we could access client.app_name.

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: Ah, didn't know that

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: Done

File: mobly/controllers/android_device.py [hunk 2]
```diff
@@ -413,9 +418,11 @@ def stop_services(self):
         if self._adb_logcat_process:
             self.stop_adb_logcat()
         self._terminate_sl4a()
-        for client in self._snippet_clients:
+        for name in list(self._snippet_clients.keys()):
+            client = self._snippet_clients[name]
```
Comment (adorokhine) [inline mobly/controllers/android_device.py]: Seems like you should be using items() here?

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: Done

File: mobly/controllers/android_device.py [hunk 3]
```diff
@@ -366,7 +370,8 @@ def __init__(self, serial=""):
         self.fastboot = fastboot.FastbootProxy(serial)
         if not self.is_bootloader and self.is_rootable:
             self.root_adb()
-        self._snippet_clients = []
+        # A dict for tracking snippet clients
```
Comment (dthkao) [inline mobly/controllers/android_device.py]: Mind adding a comment on the structure? {attr_name : client}

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: Done

File: mobly/controllers/android_device.py [hunk 4]
```diff
@@ -509,26 +514,46 @@ def root_adb(self):
...
+
+        Raises:
+            SnippetError is raised if illegal load operations are attempted.
         """
+        # Should not load snippet with the same attribute more than once.
+        if name in self._snippet_clients:
+            raise SnippetError(
+                ('Attribute "%s" is already registered with '
+                 'package "%s", it cannot be used again.') % (
+                     name, self._snippet_clients[name].app_name))
```
Comment (dthkao) [inline mobly/controllers/android_device.py]: Just realized that app_name can (and is) overwritten, which breas usage as a comparison to package. sl4a uses "SL4A".

Any thoughts on making client._package public?

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: I don't understand the concern, can you clarify

Comment (dthkao) [inline mobly/controllers/android_device.py]: Sorry this should apply to the if package==client.app_name below. Is it possible someone could reload sl4a using load_snippet()?

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: Nothing stops them from doing so, they just won't get ed started properly.
Are you saying we should guard against it?

Comment (dthkao) [inline mobly/controllers/android_device.py]: Disregard. Sl4a wouldn't load as a snippet anyway.

File: mobly/controllers/android_device.py [hunk 5]
```diff
@@ -509,26 +515,46 @@ def root_adb(self):
...
+        # Should not load snippet with an existing attribute.
+        if hasattr(self, name):
+            raise SnippetError(
+                'Attribute "%s" already exists, please use a different name.' %
+                name)
+        # Should not load the same snippet package more than once.
+        for client_name, client in self._snippet_clients.items():
+            if package == client.app_name:
```
Comment (adorokhine) [inline mobly/controllers/android_device.py]: You're making the assumption here that the snippet's app_name will always equal the package. That's not really guaranteed; it's an arbitrary string intended for user debug messages.

Do you want to expose client._package as client.package and assert on that instead?

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: I see.

Comment (xpconanfan) [inline mobly/controllers/android_device.py]: Done.
Great use case of @property :P

**Pull request merged**